containers/src/Data/Map/Internal.hs

```diff
@@ -1943,6 +1946,48 @@ withoutKeys m (Set.Bin _ k ls rs) = case splitMember k m of
 {-# INLINABLE withoutKeys #-}
 #endif
 
+-- | \(O\bigl(m \log\bigl(\frac{n}{m}+1\bigr)\bigr), \; 0 < m \leq n\). Partition the map according to a set.
+-- The first map contains the input 'Map' restricted to those keys found in the 'Set',
+-- the second map contains the input 'Map' without all keys in the 'Set'.
+-- This is more efficient than using 'restrictKeys' and 'withoutKeys' together.
+--
+-- @
+-- m \`partitionKeys\` s = (m ``restrictKeys`` s, m ``withoutKeys`` s)
+-- @
+partitionKeys :: forall k a. Ord k => Map k a -> Set k -> (Map k a, Map k a)
+partitionKeys xs ys =
+  case go xs ys of
+    xs' :*: ys' -> (xs', ys')
+  where
+    go :: Map k a -> Set k -> StrictPair (Map k a) (Map k a)
+    go Tip                 _           = Tip :*: Tip
+    go m                   Set.Tip     = Tip :*: m
+    go m@(Bin _ k x lm rm) s@Set.Bin{} =
+      case b of
+        True  -> with :*: without
+          where
+            with    =
+              if lmWith `ptrEq` lm && rmWith `ptrEq` rm
+              then m
+              else link k x lmWith rmWith
+            without =
+              link2 lmWithout rmWithout
+        False -> with :*: without
+          where
+            with    = link2 lmWith rmWith
+            without =
+              if lmWithout `ptrEq` lm && rmWithout `ptrEq` rm
+              then m
+              else link k x lmWithout rmWithout
+        where
+          !(lmWith :*: lmWithout) = go lm ls'
+          !(rmWith :*: rmWithout) = go rm rs'
+
+          !(!ls', b, !rs') = Set.splitMember k s
```

Comment:
I tried your suggestion with splitting map, was not able to measure significant difference with the current version that splits sets. They're more or less on par with each other, I can shore benchmarks if you're curious.

I like version that splits sets a little bit more because if we want to split maps then I need ta adjust `splitMember :: Ord k => k -> Map k a -> (Map k a, Bool, Map k a)` function. I'll need to lift worker out of it and having `Bool` in there is not enough - a `Maybe a` is needed instead to define `partitionKeys`. But then regular `splitMember` will need to convert `Maybe` to `Bool` which it currently doesn't do. Since everything is `INLINABLE` then Maybe may very well get allocated each time which is strictly worse that what we currently have.

Reply:
> I can share benchmarks if you're curious.

Sure, please do!

You could also use the [`SetOperations`](https://github.com/haskell/containers/blob/master/containers-tests/benchmarks/SetOperations/SetOperations.hs) setup to compare different pairs of inputs, though it will require some tweaks to work with two different types (map and set).

> ...then Maybe may very well get allocated each time which is strictly worse that what we currently have.

I don't think this necessarily indicates it would be worse. We're making O(log n) allocations with the left and right maps anyway.

Btw, `splitMember` already exists, separately implemented from `splitLookup`

https://github.com/haskell/containers/blob/c651094ec026b90c4e9d4ed81bc15eb337d3fc2e/containers/src/Data/Map/Internal.hs#L3922
https://github.com/haskell/containers/blob/c651094ec026b90c4e9d4ed81bc15eb337d3fc2e/containers/src/Data/Map/Internal.hs#L3898

Anyway, if benchmarks have shown no difference then the current implementation is fine.

Reply:
Okay, so here's my original benchmarks https://github.com/sergv/benchmark-containers. They generate random integers, turn them into strings and prefix of the list of random integers will serve as set to split on.

Here's the output, without trivial benchmarks taken from original containers.
```
All
  Map 1
    Set 0
      restrictKeys+withoutKeys:  OK
        6.86 ns ± 460 ps
      partitionKeys - split set: OK
        5.04 ns ± 376 ps, 0.73x
      partitionKeys - split map: OK
        5.05 ns ± 340 ps, 0.74x
  Map 10
    Set 1
      restrictKeys+withoutKeys:  OK
        78.8 ns ± 5.2 ns
      partitionKeys - split set: OK
        61.9 ns ± 2.7 ns
      partitionKeys - split map: OK
        68.1 ns ± 6.3 ns
    Set 9
      restrictKeys+withoutKeys:  OK
        244  ns ±  21 ns
      partitionKeys - split set: OK
        157  ns ±  13 ns, 0.64x
      partitionKeys - split map: OK
        156  ns ±  10 ns, 0.64x
  Map 100
    Set 1
      restrictKeys+withoutKeys:  OK
        98.4 ns ± 9.1 ns
      partitionKeys - split set: OK
        79.8 ns ± 5.3 ns
      partitionKeys - split map: OK
        75.8 ns ± 5.7 ns
    Set 10
      restrictKeys+withoutKeys:  OK
        1.02 μs ±  73 ns
      partitionKeys - split set: OK
        683  ns ±  60 ns
      partitionKeys - split map: OK
        677  ns ±  53 ns
    Set 99
      restrictKeys+withoutKeys:  OK
        2.21 μs ± 136 ns
      partitionKeys - split set: OK
        1.44 μs ± 124 ns, 0.65x
      partitionKeys - split map: OK
        1.14 μs ±  87 ns, 0.52x
  Map 999
    Set 1
      restrictKeys+withoutKeys:  OK
        152  ns ±  12 ns
      partitionKeys - split set: OK
        116  ns ± 6.0 ns
      partitionKeys - split map: OK
        109  ns ± 6.4 ns
    Set 10
      restrictKeys+withoutKeys:  OK
        1.66 μs ±  85 ns
      partitionKeys - split set: OK
        1.23 μs ±  86 ns
      partitionKeys - split map: OK
        1.06 μs ±  95 ns
    Set 100
      restrictKeys+withoutKeys:  OK
        9.31 μs ± 701 ns
      partitionKeys - split set: OK
        6.96 μs ± 677 ns
      partitionKeys - split map: OK
        6.36 μs ± 356 ns
    Set 998
      restrictKeys+withoutKeys:  OK
        23.8 μs ± 1.3 μs
      partitionKeys - split set: OK
        14.9 μs ± 1.3 μs, 0.63x
      partitionKeys - split map: OK
        13.5 μs ± 879 ns, 0.57x
  Map 9,988
    Set 1
      restrictKeys+withoutKeys:  OK
        167  ns ±  11 ns
      partitionKeys - split set: OK
        119  ns ±  12 ns
      partitionKeys - split map: OK
        135  ns ±  11 ns
    Set 10
      restrictKeys+withoutKeys:  OK
        1.21 μs ±  49 ns
      partitionKeys - split set: OK
        1.04 μs ±  89 ns
      partitionKeys - split map: OK
        936  ns ±  61 ns
    Set 100
      restrictKeys+withoutKeys:  OK
        14.4 μs ± 865 ns
      partitionKeys - split set: OK
        10.7 μs ± 807 ns
      partitionKeys - split map: OK
        10.2 μs ± 698 ns
    Set 1,000
      restrictKeys+withoutKeys:  OK
        131  μs ±  13 μs
      partitionKeys - split set: OK
        98.4 μs ± 6.0 μs
      partitionKeys - split map: OK
        94.3 μs ± 7.6 μs
    Set 9,987
      restrictKeys+withoutKeys:  OK
        352  μs ±  29 μs
      partitionKeys - split set: OK
        227  μs ±  18 μs, 0.64x
      partitionKeys - split map: OK
        166  μs ± 6.1 μs, 0.47x
  Map 99,237
    Set 1
      restrictKeys+withoutKeys:  OK
        252  ns ±  22 ns
      partitionKeys - split set: OK
        189  ns ±  16 ns
      partitionKeys - split map: OK
        173  ns ±  12 ns
    Set 10
      restrictKeys+withoutKeys:  OK
        2.07 μs ± 131 ns
      partitionKeys - split set: OK
        1.52 μs ± 100 ns
      partitionKeys - split map: OK
        1.62 μs ±  87 ns
    Set 100
      restrictKeys+withoutKeys:  OK
        19.8 μs ± 1.4 μs
      partitionKeys - split set: OK
        15.1 μs ± 1.4 μs
      partitionKeys - split map: OK
        14.6 μs ± 871 ns
    Set 999
      restrictKeys+withoutKeys:  OK
        251  μs ±  24 μs
      partitionKeys - split set: OK
        186  μs ±  12 μs, 0.74x
      partitionKeys - split map: OK
        179  μs ± 5.5 μs, 0.71x
    Set 9,980
      restrictKeys+withoutKeys:  OK
        2.79 ms ± 133 μs
      partitionKeys - split set: OK
        1.85 ms ±  73 μs, 0.66x
      partitionKeys - split map: OK
        1.99 ms ±  70 μs, 0.71x
    Set 99,236
      restrictKeys+withoutKeys:  OK
        4.34 ms ± 421 μs
      partitionKeys - split set: OK
        2.80 ms ± 152 μs, 0.64x
      partitionKeys - split map: OK
        2.76 ms ±  69 μs, 0.64x
  Map 989,525
    Set 1
      restrictKeys+withoutKeys:  OK
        216  ns ±  10 ns
      partitionKeys - split set: OK
        164  ns ± 6.9 ns
      partitionKeys - split map: OK
        163  ns ± 9.3 ns
    Set 10
      restrictKeys+withoutKeys:  OK
        2.31 μs ± 209 ns
      partitionKeys - split set: OK
        1.92 μs ± 173 ns
      partitionKeys - split map: OK
        1.79 μs ±  89 ns
    Set 100
      restrictKeys+withoutKeys:  OK
        22.8 μs ± 1.5 μs
      partitionKeys - split set: OK
        18.6 μs ± 1.8 μs
      partitionKeys - split map: OK
        17.7 μs ± 1.4 μs
    Set 1,000
      restrictKeys+withoutKeys:  OK
        357  μs ±  28 μs
      partitionKeys - split set: OK
        274  μs ±  14 μs
      partitionKeys - split map: OK
        273  μs ±  20 μs
    Set 9,985
      restrictKeys+withoutKeys:  OK
        9.28 ms ± 417 μs
      partitionKeys - split set: OK
        6.00 ms ± 208 μs, 0.65x
      partitionKeys - split map: OK
        7.02 ms ± 212 μs, 0.76x
    Set 99,278
      restrictKeys+withoutKeys:  OK
        84.3 ms ± 6.2 ms
      partitionKeys - split set: OK
        62.1 ms ± 4.6 ms, 0.74x
      partitionKeys - split map: OK
        60.3 ms ± 4.5 ms, 0.71x
    Set 989,524
      restrictKeys+withoutKeys:  OK
        136  ms ±  12 ms
      partitionKeys - split set: OK
        89.1 ms ± 3.6 ms, 0.65x
      partitionKeys - split map: OK
        84.8 ms ± 8.3 ms, 0.62x
```

I have also made SetOperations-like benchmark within containers, it's at https://github.com/sergv/containers/tree/benchmark-partitionKeys (previous benchmark should be run against this branch as well bacuse this is where `partitionKeysSplitMap` is defined). The results are mixed here: sometimes it looks like map splitting version is 10% faster but then I look at the error bounds and they overlap.

```
All
  partitionKeys-block_nn:               OK
    317  μs ±  29 μs
  partitionKeys-block_nn_swap:          OK
    339  μs ±  16 μs
  partitionKeys-block_ns:               OK
    37.1 μs ± 3.3 μs
  partitionKeys-block_sn_swap:          OK
    39.5 μs ± 3.0 μs
  partitionKeys-common_nn:              OK
    4.25 ms ± 196 μs
  partitionKeys-common_nn_swap:         OK
    550  μs ±  50 μs
  partitionKeys-common_ns:              OK
    1.73 ms ± 167 μs
  partitionKeys-common_nt:              OK
    83.3 μs ± 6.1 μs
  partitionKeys-common_sn_swap:         OK
    721  μs ±  70 μs
  partitionKeys-common_tn_swap:         OK
    54.8 μs ± 2.8 μs
  partitionKeys-disj_nn:                OK
    2.47 μs ± 100 ns
  partitionKeys-disj_nn_swap:           OK
    2.89 μs ± 235 ns
  partitionKeys-disj_ns:                OK
    1.99 μs ± 171 ns
  partitionKeys-disj_nt:                OK
    1.14 μs ±  64 ns
  partitionKeys-disj_sn_swap:           OK
    2.06 μs ±  92 ns
  partitionKeys-disj_tn_swap:           OK
    1.34 μs ±  93 ns
  partitionKeys-mix_nn:                 OK
    4.00 ms ± 400 μs
  partitionKeys-mix_nn_swap:            OK
    4.32 ms ± 360 μs
  partitionKeys-mix_ns:                 OK
    967  μs ±  50 μs
  partitionKeys-mix_nt:                 OK
    62.5 μs ± 4.3 μs
  partitionKeys-mix_sn_swap:            OK
    994  μs ±  56 μs
  partitionKeys-mix_tn_swap:            OK
    67.9 μs ± 6.2 μs
  partitionKeysSplitMap-block_nn:       OK
    283  μs ±  16 μs, 0.89x
  partitionKeysSplitMap-block_nn_swap:  OK
    298  μs ±  22 μs, 0.88x
  partitionKeysSplitMap-block_ns:       OK
    30.0 μs ± 1.3 μs, 0.81x
  partitionKeysSplitMap-block_sn_swap:  OK
    28.3 μs ± 1.3 μs, 0.72x
  partitionKeysSplitMap-common_nn:      OK
    4.56 ms ± 352 μs, 1.07x
  partitionKeysSplitMap-common_nn_swap: OK
    428  μs ±  16 μs, 0.78x
  partitionKeysSplitMap-common_ns:      OK
    2.75 ms ±  54 μs, 1.59x
  partitionKeysSplitMap-common_nt:      OK
    85.4 μs ± 5.9 μs, 1.03x
  partitionKeysSplitMap-common_sn_swap: OK
    941  μs ±  58 μs, 1.31x
  partitionKeysSplitMap-common_tn_swap: OK
    55.1 μs ± 3.0 μs, 1.00x
  partitionKeysSplitMap-disj_nn:        OK
    2.03 μs ± 196 ns, 0.82x
  partitionKeysSplitMap-disj_nn_swap:   OK
    1.83 μs ± 166 ns, 0.63x
  partitionKeysSplitMap-disj_ns:        OK
    1.51 μs ± 103 ns, 0.76x
  partitionKeysSplitMap-disj_nt:        OK
    946  ns ±  52 ns, 0.83x
  partitionKeysSplitMap-disj_sn_swap:   OK
    1.51 μs ±  94 ns, 0.73x
  partitionKeysSplitMap-disj_tn_swap:   OK
    896  ns ±  43 ns, 0.67x
  partitionKeysSplitMap-mix_nn:         OK
    4.40 ms ± 199 μs, 1.10x
  partitionKeysSplitMap-mix_nn_swap:    OK
    3.50 ms ± 155 μs, 0.81x
  partitionKeysSplitMap-mix_ns:         OK
    832  μs ±  57 μs, 0.86x
  partitionKeysSplitMap-mix_nt:         OK
    67.4 μs ± 5.5 μs, 1.08x
  partitionKeysSplitMap-mix_sn_swap:    OK
    873  μs ±  82 μs, 0.88x
  partitionKeysSplitMap-mix_tn_swap:    OK
    57.1 μs ± 4.0 μs, 0.84x

All 44 tests passed (48.06s)
```

Reply:
Thanks! So I think we can conclude that one is not definitively better than the other.